### ISSUE TITLE
psycopg2 connection can be left idle in transaction

### ISSUE THREAD

**Issue by andy-d** [pydata/pandas#11521]
repro:
(1) make a psycopg2 conn
(2) make a pandas df
(3) call df.read_sql_query("SELECT pg_sleep(1000);", conn)
(4) kill the terminal

actual: connection on backend is left hanging as idle in transaction

**Comment by Ricky811** [comment]
I have get the same problem when i use pd.read_query() , I not  cloas the connection.  in my database  the connection state is `idle in transaction`. How can I make the connection idle?

**Comment by fukac99** [comment]
any progress on this? I have the same problem

**Comment by TomAugspurger** [comment]
Why do we think this is a pandas issue? When psycopg2 is used alone, does it correctly terminate the connection when the process is kliled?

**Comment by maxim-lian** [comment]
Or even a python issue? If you 'kill the power', how would the transaction be closed? 

(not saying this isn't an issue, but need an example where _pandas_ is failing to close a connection)

**Comment by TomAugspurger** [comment]
Indeed. In theory, psycopg2 could catch SIGINTs and try to abort the transaction. I'm not sure what's actually done.

**Comment by TomAugspurger** [comment]
The first item in their FAQ is suggestive: http://initd.org/psycopg/docs/faq.html#problems-with-transactions-handling

**Comment by AlbertRothman** [comment]
I found that pd.read_sql_query("SELECT * from mytable;", **conn**) leaves the transaction hanging as as "idle in transaction", but using  pd.read_sql_query("SELECT * from mytable;", **engine**) closes the transaction with a ROLLBACK. I was using SQLAlchemy as my engine.

**Comment by mroeschke** [comment]
It's not clear this is a pandas issue specifically so closing.
